leader-election-impl/src/main/java/com/palantir/paxos/SplittingPaxosStateLog.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.paxos;
+
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.immutables.value.Value;
+
+import com.palantir.common.persist.Persistable;
+import com.palantir.logsafe.Preconditions;
+
+/**
+ * This implementation of {@link PaxosStateLog} delegates all reads and writes of rounds to one of two delegates, as
+ * determined by the cutoff point. If a read or write does occur prior to the cutoff point, i.e., to the legacy delegate
+ * we update the appropriate metric. Remaining methods are delegated only to the current delegate.
+ */
+public final class SplittingPaxosStateLog<V extends Persistable & Versionable> implements PaxosStateLog<V> {
+    private final PaxosStateLog<V> legacyLog;
+    private final PaxosStateLog<V> currentLog;
+    private final Runnable markLegacyWrite;
+    private final Runnable markLegacyRead;
+    private final long cutoffInclusive;
+    private final AtomicLong leastLogEntry;
+
+    private SplittingPaxosStateLog(PaxosStateLog<V> legacyLog,
+            PaxosStateLog<V> currentLog,
+            Runnable markLegacyWrite,
+            Runnable markLegacyRead,
+            long cutoffInclusive,
+            AtomicLong leastLogEntry) {
+        this.legacyLog = legacyLog;
+        this.currentLog = currentLog;
+        this.markLegacyWrite = markLegacyWrite;
+        this.markLegacyRead = markLegacyRead;
+        this.cutoffInclusive = cutoffInclusive;
+        this.leastLogEntry = leastLogEntry;
+    }
+
+    public static <V extends Persistable & Versionable> PaxosStateLog<V> create(SplittingParameters<V> parameters) {
+        return new SplittingPaxosStateLog<>(
+                parameters.legacyLog(),
+                parameters.currentLog(),
+                parameters.markLegacyWrite(),
+                parameters.markLegacyRead(),
+                parameters.cutoffInclusive(),
+                new AtomicLong(Math.min(parameters.legacyLog().getLeastLogEntry(), parameters.cutoffInclusive())));
```

Comment:
Discussed offline: This is ok as long as we make sure internal server doesn't say that it is fully ready until this has completed for all clients.

Reply:
We can also hack in an optimisation to make `getLeastLogEntry()` faster at a majority of installations.